src/libopensc/card-piv.c

```diff
@@ -2992,38 +3056,92 @@ static int piv_match_card(sc_card_t *card)
 	if (card->type == -1)
 		card->type = type;
 
-	card->drv_data = priv;
+	card->drv_data = priv; /* will free if no match, or pass on to piv_init */
 	priv->aid_file = sc_file_new();
 	priv->selected_obj = -1;
 	priv->pin_preference = 0x80; /* 800-73-3 part 1, table 3 */
 	priv->logged_in = SC_PIN_STATE_UNKNOWN;
 	priv->tries_left = 10; /* will assume OK at start */
+	priv->pstate = PIV_STATE_MATCH;
 
 	/* Some objects will only be present if Histroy object says so */
 	for (i=0; i < PIV_OBJ_LAST_ENUM -1; i++)
 		if(piv_objects[i].flags & PIV_OBJECT_NOT_PRESENT)
 			priv->obj_cache[i].flags |= PIV_OBJ_CACHE_NOT_PRESENT;
 
+	sc_lock(card);
 
+	/*
+	 * detect if active AID is PIV. NIST 800-73 says Only one PIV application per card
+	 * and PIV must be the default application
+	 * This can avoid doing doing a select_aid and losing the login state on some cards
+	 * We may get interference on some cards by other drivers trying SELECT_AID before
+	 * we get to see if PIV application is still active.
+	 * putting PIV driver first might help. 
+	 * This may fail if the wrong AID is active
+	 */
+	i = piv_find_discovery(card);
+
+	if (i < 0) {
+		/* Detect by selecting applet */
+		i = piv_find_aid(card, &aidfile);
+	}
+
+	if (i >= 0) {
+		/*
+		 * We now know PIV AID is active, test DISCOVERY object 
+		 * Some CAC cards with PIV don't support DISCOVERY and return 
+		 * SC_ERROR_INCORRECT_PARAMETERS. Any error other then 
+		 * SC_ERROR_FILE_NOT_FOUND means we cannot use discovery 
+		 * to test for active AID.
+		 */
+		i7e = piv_find_discovery(card);
+		if (i7e != 0 && i7e !=  SC_ERROR_FILE_NOT_FOUND) {
+			priv->card_issues |= CI_DISCOVERY_USELESS;
+			priv->obj_cache[PIV_OBJ_DISCOVERY].flags |= PIV_OBJ_CACHE_NOT_PRESENT;
+		}
+	}
+
+
+	if (i < 0) {
+		piv_finish(card);
+		/* don't match. Does not have a PIV applet. */
+		sc_unlock(card);
+		return 0;
+	}
+
+	/* Matched, and priv is being passed to piv_init */
+	/* hold the lock and pass to piv_init */
+	priv->pstate=PIV_STATE_INIT;
 	return 1; /* match */
 }
 
 
 static int piv_init(sc_card_t *card)
 {
-	int r;
-	piv_private_data_t * priv = PIV_DATA(card);
+	int r = 0;
+	piv_private_data_t * priv = NULL;
 	sc_apdu_t apdu;
 	unsigned long flags;
 	unsigned long ext_flags;
 	u8 neo_version_buf[3];
```

Comment:
This is separate unrelated issue, but the `neo_version` should have different name since it is used also for getting a version of Yubikey 4.